File: bindings/cpp/src/bignum.cpp
// bignum.cpp - the magnitude bignum over uint64_t limbs.

#include "bignum.hpp"

#include <algorithm>
#include <stdexcept>

namespace co {

int big_cmp(const Big& a, const Big& b) {
    if (a.l.size() != b.l.size())
        return a.l.size() < b.l.size() ? -1 : 1;
    for (size_t i = a.l.size(); i-- > 0;) {
        if (a.l[i] != b.l[i]) return a.l[i] < b.l[i] ? -1 : 1;
    }
    return 0;
}

Big big_from_u64(uint64_t v) {
    Big r;
    if (v) r.l.push_back(v);
    return r;
}

Big big_add(const Big& a, const Big& b) {
    Big r;
    size_t n = std::max(a.l.size(), b.l.size());
    r.l.resize(n + 1, 0);
    unsigned __int128 carry = 0;
    for (size_t i = 0; i < n; ++i) {
        unsigned __int128 s = carry;
        if (i < a.l.size()) s += a.l[i];
        if (i < b.l.size()) s += b.l[i];
        r.l[i] = static_cast<uint64_t>(s);
        carry = s >> 64;
    }
    r.l[n] = static_cast<uint64_t>(carry);
    r.norm();
    return r;
}

Big big_sub(const Big& a, const Big& b) {
    if (big_cmp(a, b) < 0)
        throw std::runtime_error("big_sub: negative result (magnitude only)");
    Big r;
    r.l.assign(a.l.size(), 0);
    uint64_t borrow = 0;
    for (size_t i = 0; i < a.l.size(); ++i) {
        uint64_t bi = i < b.l.size() ? b.l[i] : 0;
        unsigned __int128 lhs = a.l[i];
        unsigned __int128 rhs = static_cast<unsigned __int128>(bi) + borrow;
        if (lhs >= rhs) {
            r.l[i] = static_cast<uint64_t>(lhs - rhs);
            borrow = 0;
        } else {
            r.l[i] = static_cast<uint64_t>(
                (static_cast<unsigned __int128>(1) << 64) + lhs - rhs);
            borrow = 1;
        }
    }
    r.norm();
    return r;
}

Big big_mul(const Big& a, const Big& b) {
    Big r;
    if (a.isZero() || b.isZero()) return r;
    r.l.assign(a.l.size() + b.l.size(), 0);
    for (size_t i = 0; i < a.l.size(); ++i) {
        unsigned __int128 carry = 0;
        for (size_t j = 0; j < b.l.size(); ++j) {
            unsigned __int128 cur =
                static_cast<unsigned __int128>(a.l[i]) * b.l[j] +
                r.l[i + j] + carry;
            r.l[i + j] = static_cast<uint64_t>(cur);
            carry = cur >> 64;
        }
        size_t k = i + b.l.size();
        while (carry) {
            unsigned __int128 cur = static_cast<unsigned __int128>(r.l[k]) + carry;
            r.l[k] = static_cast<uint64_t>(cur);
            carry = cur >> 64;
            ++k;
        }
    }
    r.norm();
    return r;
}

Big big_shl(const Big& a, size_t bits) {
    if (a.isZero()) return a;
    size_t limbShift = bits / 64, bitShift = bits % 64;
    Big r;
    r.l.assign(a.l.size() + limbShift + 1, 0);
    for (size_t i = 0; i < a.l.size(); ++i) {
        r.l[i + limbShift] |= bitShift ? (a.l[i] << bitShift) : a.l[i];
        if (bitShift)
            r.l[i + limbShift + 1] |= a.l[i] >> (64 - bitShift);
    }
    r.norm();
    return r;
}

Big big_shr(const Big& a, size_t bits) {
    size_t limbShift = bits / 64, bitShift = bits % 64;
    Big r;
    if (limbShift >= a.l.size()) return r;
    r.l.assign(a.l.size() - limbShift, 0);
    for (size_t i = 0; i < r.l.size(); ++i) {
        uint64_t lo = a.l[i + limbShift];
        uint64_t hi = (i + limbShift + 1 < a.l.size()) ? a.l[i + limbShift + 1] : 0;
        r.l[i] = bitShift ? ((lo >> bitShift) | (hi << (64 - bitShift))) : lo;
    }
    r.norm();
    return r;
}

Big big_lowbits(const Big& a, size_t bits) {
    size_t limbCount = (bits + 63) / 64;
    Big r;
    r.l.assign(a.l.begin(),
               a.l.begin() + static_cast<long>(std::min(limbCount, a.l.size())));
    size_t rem = bits % 64;
    if (rem && r.l.size() == limbCount)
        r.l.back() &= (uint64_t(1) << rem) - 1;
    r.norm();
    return r;
}

size_t big_bitlen(const Big& a) {
    if (a.isZero()) return 0;
    uint64_t top = a.l.back();
    size_t bits = (a.l.size() - 1) * 64;
    while (top) {
        ++bits;
        top >>= 1;
    }
    return bits;
}

bool big_bit(const Big& a, size_t i) {
    size_t limb = i / 64;
    if (limb >= a.l.size()) return false;
    return (a.l[limb] >> (i % 64)) & 1;
}

Big big_mod(const Big& a, const Big& m) {
    if (m.isZero()) throw std::runtime_error("big_mod: modulus is zero");
    if (big_cmp(a, m) < 0) return a;
    // Shift-subtract long division: align m to a's magnitude, walk down.
    size_t shift = big_bitlen(a) - big_bitlen(m);
    Big r = a;
    Big t = big_shl(m, shift);
    for (size_t i = 0; i <= shift; ++i) {
        if (big_cmp(r, t) >= 0) r = big_sub(r, t);
        t = big_shr(t, 1);
    }
    return r;
}

Big big_modpow(const Big& b, const Big& e, const Big& m) {
    Big result = big_from_u64(1);
    result = big_mod(result, m);
    Big base = big_mod(b, m);
    size_t bits = big_bitlen(e);
    // Left-to-right square-and-multiply.
    for (size_t i = bits; i-- > 0;) {
        result = big_mod(big_mul(result, result), m);
        if (big_bit(e, i)) result = big_mod(big_mul(result, base), m);
    }
    return result;
}

Big big_from_bytes_le(const std::string& bytes) {
    Big r;
    r.l.assign((bytes.size() + 7) / 8, 0);
    for (size_t i = 0; i < bytes.size(); ++i)
        r.l[i / 8] |= static_cast<uint64_t>(
                          static_cast<unsigned char>(bytes[i]))
                      << (8 * (i % 8));
    r.norm();
    return r;
}

std::string big_to_bytes_le(const Big& a, size_t n) {
    if (big_bitlen(a) > n * 8)
        throw std::runtime_error("big_to_bytes_le: value too large");
    std::string out(n, '\0');
    for (size_t i = 0; i < n; ++i) {
        size_t limb = i / 8;
        if (limb < a.l.size())
            out[i] = static_cast<char>((a.l[limb] >> (8 * (i % 8))) & 0xff);
    }
    return out;
}

Big big_from_hex(const std::string& hex) {
    Big r;
    for (char c : hex) {
        int d;
        if (c >= '0' && c <= '9') d = c - '0';
        else if (c >= 'a' && c <= 'f') d = c - 'a' + 10;
        else if (c >= 'A' && c <= 'F') d = c - 'A' + 10;
        else throw std::runtime_error("big_from_hex: bad digit");
        r = big_add(big_shl(r, 4), big_from_u64(static_cast<uint64_t>(d)));
    }
    return r;
}

std::string big_to_hex(const Big& a) {
    if (a.isZero()) return "0";
    static const char* digits = "0123456789abcdef";
    std::string out;
    for (size_t i = a.l.size(); i-- > 0;) {
        for (int nib = 15; nib >= 0; --nib) {
            int d = static_cast<int>((a.l[i] >> (4 * nib)) & 0xf);
            if (out.empty() && d == 0) continue;
            out.push_back(digits[d]);
        }
    }
    if (out.empty()) out = "0";
    return out;
}

}  // namespace co

File: bindings/kotlin/src/Bignum.kt
// Arbitrary-precision non-negative integers for Ed25519 and exact JCS integers.
//
// Representation: IntArray of base-2^16 limbs, little-endian, normalized (no
// leading zero limbs; zero is the empty array). Base 2^16 keeps every Long
// product safely inside 64 bits: 32 limbs x 32 limbs accumulates at most
// 32 * (2^16-1)^2 < 2^37 per column, far from overflow.
package org.causalontology

object Bignum {

    val ZERO = IntArray(0)
    val ONE = intArrayOf(1)

    // Trim leading (high-order) zero limbs so every value has one canonical form.
    fun norm(a: IntArray): IntArray {
        var n = a.size
        while (n > 0 && a[n - 1] == 0) n--
        return if (n == a.size) a else a.copyOf(n)
    }

    fun isZero(a: IntArray): Boolean = a.isEmpty()

    fun fromLong(v: Long): IntArray {
        require(v >= 0) { "Bignum is non-negative" }
        var x = v
        val out = mutableListOf<Int>()
        while (x != 0L) {
            out.add((x and 0xFFFF).toInt())
            x = x ushr 16
        }
        return out.toIntArray()
    }

    // Parse a big-endian hexadecimal string (no sign, no prefix).
    fun fromHex(hex: String): IntArray {
        var acc = ZERO
        val sixteen = intArrayOf(16)
        for (c in hex) {
            val d = when (c) {
                in '0'..'9' -> c - '0'
                in 'a'..'f' -> c - 'a' + 10
                in 'A'..'F' -> c - 'A' + 10
                else -> throw IllegalArgumentException("bad hex digit $c")
            }
            acc = add(mul(acc, sixteen), fromLong(d.toLong()))
        }
        return acc
    }

    // Little-endian byte decoding (the Ed25519 wire order).
    fun fromBytesLE(b: ByteArray): IntArray {
        val limbs = IntArray((b.size + 1) / 2)
        for (i in b.indices) {
            val v = b[i].toInt() and 0xFF
            if (i % 2 == 0) limbs[i / 2] = limbs[i / 2] or v
            else limbs[i / 2] = limbs[i / 2] or (v shl 8)
        }
        return norm(limbs)
    }

    // Little-endian byte encoding, zero-padded to exactly len bytes.
    fun toBytesLE(a: IntArray, len: Int): ByteArray {
        val out = ByteArray(len)
        for (i in 0 until len) {
            val limb = i / 2
            if (limb >= a.size) break
            val v = if (i % 2 == 0) a[limb] and 0xFF else (a[limb] ushr 8) and 0xFF
            out[i] = v.toByte()
        }
        return out
    }

    fun cmp(a: IntArray, b: IntArray): Int {
        if (a.size != b.size) return if (a.size > b.size) 1 else -1
        for (i in a.size - 1 downTo 0) {
            if (a[i] != b[i]) return if (a[i] > b[i]) 1 else -1
        }
        return 0
    }

    fun add(a: IntArray, b: IntArray): IntArray {
        val n = maxOf(a.size, b.size)
        val out = IntArray(n + 1)
        var carry = 0
        for (i in 0 until n) {
            val s = (if (i < a.size) a[i] else 0) + (if (i < b.size) b[i] else 0) + carry
            out[i] = s and 0xFFFF
            carry = s ushr 16
        }
        out[n] = carry
        return norm(out)
    }

    // Subtraction; requires a >= b (the caller keeps everything non-negative).
    fun sub(a: IntArray, b: IntArray): IntArray {
        require(cmp(a, b) >= 0) { "Bignum.sub underflow" }
        val out = IntArray(a.size)
        var borrow = 0
        for (i in a.indices) {
            var s = a[i] - (if (i < b.size) b[i] else 0) - borrow
            if (s < 0) { s += 0x10000; borrow = 1 } else borrow = 0
            out[i] = s
        }
        return norm(out)
    }

    // Schoolbook multiplication with a Long accumulator per column.
    fun mul(a: IntArray, b: IntArray): IntArray {
        if (a.isEmpty() || b.isEmpty()) return ZERO
        val acc = LongArray(a.size + b.size)
        for (i in a.indices) {
            val ai = a[i].toLong()
            for (j in b.indices) {
                acc[i + j] += ai * b[j]
            }
            // Propagate partial carries every row so columns never overflow Long.
            if (i % 16 == 15) {
                var carry = 0L
                for (k in acc.indices) {
                    val t = acc[k] + carry
                    acc[k] = t and 0xFFFF
                    carry = t ushr 16
                }
            }
        }
        var carry = 0L
        val out = IntArray(acc.size)
        for (k in acc.indices) {
            val t = acc[k] + carry
            out[k] = (t and 0xFFFF).toInt()
            carry = t ushr 16
        }
        require(carry == 0L)
        return norm(out)
    }

    fun bitLen(a: IntArray): Int {
        if (a.isEmpty()) return 0
        var top = a[a.size - 1]
        var bits = (a.size - 1) * 16
        while (top != 0) { bits++; top = top ushr 1 }
        return bits
    }

    fun testBit(a: IntArray, i: Int): Boolean {
        val limb = i / 16
        if (limb >= a.size) return false
        return (a[limb] ushr (i % 16)) and 1 == 1
    }

    fun shl(a: IntArray, bits: Int): IntArray {
        if (a.isEmpty() || bits == 0) return a
        val limbShift = bits / 16
        val bitShift = bits % 16
        val out = IntArray(a.size + limbShift + 1)
        for (i in a.indices) {
            val v = a[i] shl bitShift
            out[i + limbShift] = out[i + limbShift] or (v and 0xFFFF)
            out[i + limbShift + 1] = out[i + limbShift + 1] or (v ushr 16)
        }
        return norm(out)
    }

    fun shr(a: IntArray, bits: Int): IntArray {
        if (a.isEmpty() || bits == 0) return a
        val limbShift = bits / 16
        val bitShift = bits % 16
        if (limbShift >= a.size) return ZERO
        val out = IntArray(a.size - limbShift)
        for (i in out.indices) {
            var v = a[i + limbShift] ushr bitShift
            if (bitShift != 0 && i + limbShift + 1 < a.size) {
                v = v or ((a[i + limbShift + 1] shl (16 - bitShift)) and 0xFFFF)
            }
            out[i] = v
        }
        return norm(out)
    }

    // General modulus by binary shift-and-subtract (used for the group order q;
    // the field prime p has its own fast fold in Ed25519).
    fun mod(x: IntArray, m: IntArray): IntArray {
        require(!isZero(m)) { "modulus is zero" }
        if (cmp(x, m) < 0) return x
        var r = x
        var shift = bitLen(x) - bitLen(m)
        var t = shl(m, shift)
        while (shift >= 0) {
            if (cmp(r, t) >= 0) r = sub(r, t)
            t = shr(t, 1)
            shift--
        }
        return r
    }

    // Divide by a small positive divisor; returns quotient and remainder.
    fun divSmall(a: IntArray, d: Int): Pair<IntArray, Int> {
        require(d in 1..0xFFFF)
        val out = IntArray(a.size)
        var rem = 0L
        for (i in a.size - 1 downTo 0) {
            val cur = (rem shl 16) or a[i].toLong()
            out[i] = (cur / d).toInt()
            rem = cur % d
        }
        return Pair(norm(out), rem.toInt())
    }

    // Exact decimal rendering (used by JCS for integral doubles beyond 2^63).
    fun toDecimalString(a: IntArray): String {
        if (isZero(a)) return "0"
        val sb = StringBuilder()
        var cur = a
        while (!isZero(cur)) {
            val (q, r) = divSmall(cur, 10000)
            cur = q
            if (isZero(cur)) sb.append(r.toString().reversed())
            else sb.append(r.toString().padStart(4, '0').reversed())
        }
        return sb.toString().reversed()
    }
}
